### ISSUE TITLE
Hanging Line on Docker Build on Windows and OSX

### ISSUE THREAD

**Issue by max-hanrahan** [e-mission/em-public-dashboard#37]
The hanging line is: 

`RUN /bin/bash -c "cd e-mission-server && source setup/activate.sh && conda env update --name emission --file setup/environment36.notebook.additions.yml"`

Reproduced using docker containers and Anaconda environment on my home computer.

**Comment by max-hanrahan** [comment]
When I run that line I receive the following:
`CondaValueError: could not parse 'name: emission' in: C:\Users\mhanr\Desktop\e-mission-server\setup\environment36.notebook.additions.yml`

**Comment by shankari** [comment]
Adding a `tail -f` to pause execution before the conda commands so that we can run them manually. Idea from 
https://github.com/e-mission/e-mission-docker/blob/master/clone_server.sh

```
$ git diff viz_scripts/docker/
diff --git a/viz_scripts/docker/Dockerfile b/viz_scripts/docker/Dockerfile
index 4604044..48560e5 100644
--- a/viz_scripts/docker/Dockerfile
+++ b/viz_scripts/docker/Dockerfile
@@ -3,8 +3,8 @@ FROM emission/e-mission-server:2.9.2

 ADD environment36.dashboard.additions.yml /

-RUN /bin/bash -c "cd e-mission-server && source setup/activate.sh && conda env update --name emission --file setup/environment36.notebook.additions.yml"
-RUN /bin/bash -c "cd e-mission-server && source setup/activate.sh && conda env update --name emission --file /environment36.dashboard.additions.yml"
+# RUN /bin/bash -c "cd e-mission-server && source setup/activate.sh && conda env update --name emission --file setup/environment36.notebook.additions.yml"
+# RUN /bin/bash -c "cd e-mission-server && source setup/activate.sh && conda env update --name emission --file /environment36.dashboard.additions.yml"

 ADD start_notebook.sh /usr/src/app/start_notebook.sh
 RUN chmod u+x /usr/src/app/start_notebook.sh
diff --git a/viz_scripts/docker/start_notebook.sh b/viz_scripts/docker/start_notebook.sh
index d08a5e9..6c422bc 100755
--- a/viz_scripts/docker/start_notebook.sh
+++ b/viz_scripts/docker/start_notebook.sh
@@ -1,6 +1,10 @@
 #!/usr/bin/env bash
 #Configure web server

+# # use this as the launch script instead if you want to experiment with docker
+# build changes
+tail -f /environment36.dashboard.additions.yml
+
 pushd /usr/src/app/e-mission-server

 #set database URL using environment variable
```

**Comment by shankari** [comment]
Ran `cd e-mission-server && source setup/activate.sh && conda env update --name emission --file setup/environment36.notebook.additions.yml` and it hung again.

Let's try to install the packages one by one

**Comment by shankari** [comment]
### Branca: failed

```
# conda install branca=0.4.1
Collecting package metadata (current_repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Collecting package metadata (repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.

PackagesNotFoundError: The following packages are not available from current channels:

  - branca=0.4.1
```

### folium: failed

```
conda install folium=0.11.0
Collecting package metadata (current_repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Collecting package metadata (repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.

PackagesNotFoundError: The following packages are not available from current channels:

  - folium=0.11.0
```

### ipython: works

```
# conda install ipython=7.17.0
Collecting package metadata (current_repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Collecting package metadata (repodata.json): done
Solving environment: done
```

### jedi: works


```
# conda install jedi=0.17.2
Collecting package metadata (current_repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: failed with repodata from current_repodata.json, will retry with next repodata source.
Collecting package metadata (repodata.json): done
```

### jupyter: works

```
# conda install jupyter=1.0.0
Collecting package metadata (current_repodata.json): done
Solving environment: done

```

### matplotlib:  works

```
# conda install matplotlib=3.2.2
Collecting package metadata (current_repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Collecting package metadata (repodata.json): done
Solving environment: done
```

We should really upgrade everything since the frozen solve seems to fail pretty consistently

**Comment by shankari** [comment]
Upgraded everything to the latest versions and failed

```
# source setup/activate.sh && conda env update --name emission --file setup/environment36.notebook.additions.yml
Collecting package metadata (repodata.json): done
Solving environment: |
Found conflicts! Looking for incompatible packages.
This can take several minutes.  Press CTRL-C to abort.
failed                                                                                                     \
Solving environment: |
Found conflicts! Looking for incompatible packages.
This can take several minutes.  Press CTRL-C to abort.
failed                                                                                                     \

UnsatisfiableError: The following specifications were found
to be incompatible with the existing python installation in your environment:

Specifications:

  - ipython=8.4.0 -> python[version='>=3.10,<3.11.0a0|>=3.8,<3.9.0a0|>=3.9,<3.10.0a0']

Your python: python=3.7

```

<details>
<summary>And a bunch of other conflicts</summary>

```
The following specifications were found to be incompatible with each other:

Output in format: Requested package -> Available versions

Package numpy conflicts for:
matplotlib=3.5.2 -> matplotlib-base[version='>=3.5.2,<3.5.3.0a0'] -> numpy[version='>=1.17|>=1.19.5,<2.0a0|>=1.21.6,<2.0a0']
folium=0.12.0 -> numpy

Package gdbm conflicts for:
jedi=0.18.1 -> pypy3.8[version='>=7.3.8'] -> gdbm[version='>=1.18,<1.19.0a0']
jupyter=1.0.0 -> pypy3.7[version='>=7.3.7'] -> gdbm[version='>=1.18,<1.19.0a0']
python=3.7 -> pypy3.7=7.3.7 -> gdbm[version='>=1.18,<1.19.0a0']
matplotlib=3.5.2 -> pypy3.7[version='>=7.3.7'] -> gdbm[version='>=1.18,<1.19.0a0']

Package pypy3.8 conflicts for:
jupyter=1.0.0 -> python[version='>=3.8,<3.9.0a0'] -> pypy3.8[version='7.3.8.*|7.3.9.*']
folium=0.12.0 -> numpy -> pypy3.8[version='7.3.8.*|7.3.9.*|>=7.3.8|>=7.3.9']
matplotlib=3.5.2 -> tornado -> pypy3.8[version='7.3.8.*|7.3.9.*|>=7.3.8']
ipython=8.4.0 -> jedi[version='>=0.16'] -> pypy3.8[version='7.3.8.*|7.3.9.*|>=7.3.8|>=7.3.9']
branca=0.5.0 -> python[version='>=3'] -> pypy3.8[version='7.3.8.*|7.3.9.*|>=7.3.9|>=7.3.8']
jedi=0.18.1 -> python[version='>=3.8,<3.9.0a0'] -> pypy3.8[version='7.3.*|7.3.8.*|7.3.9.*']
jedi=0.18.1 -> pypy3.8[version='>=7.3.8']

Package ca-certificates conflicts for:
python=3.7 -> openssl[version='>=1.1.1l,<1.1.2a'] -> ca-certificates
jupyter=1.0.0 -> python -> ca-certificates

Package certifi conflicts for:
branca=0.5.0 -> setuptools -> certifi[version='>=2016.09|>=2016.9.26']
ipython=8.4.0 -> setuptools[version='>=18.5'] -> certifi[version='>=2016.09|>=2016.9.26']
folium=0.12.0 -> requests -> certifi[version='>=2017.4.17']
matplotlib=3.5.2 -> matplotlib-base[version='>=3.5.2,<3.5.3.0a0'] -> certifi[version='>=2020.06.20']

Package jinja2 conflicts for:
folium=0.12.0 -> branca[version='>=0.3.0'] -> jinja2
folium=0.12.0 -> jinja2[version='>=2.9']

Package pypy3.6 conflicts for:
matplotlib=3.5.2 -> tornado -> pypy3.6[version='>=7.3.1|>=7.3.2|>=7.3.3']
jupyter=1.0.0 -> pypy3.6[version='>=7.3.3']
jupyter=1.0.0 -> ipykernel -> pypy3.6[version='7.3.*|7.3.0.*|7.3.1.*|7.3.2.*|7.3.3.*|>=7.3.1|>=7.3.2']
ipython=8.4.0 -> jedi[version='>=0.16'] -> pypy3.6[version='>=7.3.1|>=7.3.2|>=7.3.3']
branca=0.5.0 -> python[version='>=3'] -> pypy3.6[version='7.3.0.*|7.3.1.*|7.3.2.*|7.3.3.*|>=7.3.3|>=7.3.2|>=7.3.1']
folium=0.12.0 -> numpy -> pypy3.6[version='7.3.0.*|7.3.1.*|7.3.2.*|7.3.3.*|>=7.3.1|>=7.3.2|>=7.3.3']

Package six conflicts for:
folium=0.12.0 -> branca[version='>=0.3.0'] -> six
branca=0.5.0 -> six
ipython=8.4.0 -> prompt-toolkit[version='>=2.0.0,<3.1.0,!=3.0.0,!=3.0.1'] -> six[version='>=1.9.0']

Package packaging conflicts for:
jupyter=1.0.0 -> ipykernel -> packaging
matplotlib=3.5.2 -> matplotlib-base[version='>=3.5.2,<3.5.3.0a0'] -> packaging[version='>=20.0']

Package markupsafe conflicts for:
jupyter=1.0.0 -> nbconvert -> markupsafe[version='>=2.0']
branca=0.5.0 -> jinja2 -> markupsafe[version='>=0.23|>=0.23,<2|>=2.0|>=2.0.0rc2']
folium=0.12.0 -> jinja2[version='>=2.9'] -> markupsafe[version='>=0.23|>=0.23,<2|>=2.0|>=2.0.0rc2']

Package setuptools conflicts for:
branca=0.5.0 -> setuptools
jupyter=1.0.0 -> ipykernel -> setuptools[version='>=60']
ipython=8.4.0 -> pygments[version='>=2.4.0'] -> setuptools
ipython=8.4.0 -> setuptools[version='>=18.5']
python=3.7 -> pip -> setuptools
folium=0.12.0 -> branca[version='>=0.3.0'] -> setuptools
```
</details>

A lot of these conflicts seem to be between dependencies that have versions and dependencies that don't
Would it help to manually list the dependencies (with concrete values) so the solver wouldn't have to do too much work?

**Comment by shankari** [comment]
Removing ipython since jupyter doesn't need it and nobody uses the command shell otherwise other than me

**Comment by shankari** [comment]
Let's try to upgrade the base image and miniconda and see if that helps...

**Comment by shankari** [comment]
Since this is taking a long time and I have a lot of tasks, decided to fall back to using a pre-created image from the time that this did work.

**Comment by shankari** [comment]
These are the changes to experiment with upgrading the image.
[upgrade_experiment.patch.gz](https://github.com/e-mission/em-public-dashboard/files/8865557/upgrade_experiment.patch.gz)

**Comment by shankari** [comment]
@aGuttman can you handle finding and pinning the correct versions of the packages so that docker image builds again without using a pinned image? The pinned image is likely on an older version of ubuntu and might not meet cloud services standards for deployment.

This has a hard deadline of next Tuesday, since we aim to get the public dashboard working on the staging by the end of next week.

**Comment by aGuttman** [comment]
`conda 4.8.3` - `4.12.0` will solve and install the packages correctly with these versions:

```
name: emission
channels:
- conda-forge
- defaults
dependencies:
- branca=0.5.0
- folium=0.12.1.post1
- ipython=7.33.0
- jedi=0.18.1
- jupyter=1.0.0
- matplotlib=3.5.1
```

`ipython` and `matplotlib` are not the latest versions, which are `8.4.0` and `3.5.3` respectfully.

`conda 4.12.0` is not the latest version of `conda`. There are two more recent releases, `4.13.0` and `4.14.0`, but they break our conda install script ([e-mission-server/setup/setup_conda.sh](https://github.com/e-mission/e-mission-server/blob/07e01eab5accb9b4f8204ecab86c0c79caf3dfa9/setup/setup_conda.sh)) somehow. I've not spent a lot of time investigating.
<details>
<summary>err msg</summary>

  ```
...
> [ 5/10] RUN /bin/bash -c "cd e-mission-server && cat setup/export_versions.sh && source setup/setup_conda.sh Linux-x86_64 && source setup/setup.sh":         
#10 0.217 export EXP_CONDA_VER=4.14.0Installing for version 4.14.0 and platform Linux-x86_64                                                                    
#10 0.228   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
#10 0.228                                  Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100   383  100   383    0     0    892      0 --:--:-- --:--:-- --:--:--   892
#10 0.664 miniconda.sh: line 1: syntax error near unexpected token `newline'
#10 0.664 miniconda.sh: line 1: `<html>'
#10 0.664 setup/setup_conda.sh: line 17: /root/miniconda-4.14.0/etc/profile.d/conda.sh: No such file or directory
#10 0.664 setup/setup_conda.sh: line 19: conda: command not found
#10 0.665 setup/setup_conda.sh: line 21: conda: command not found
#10 0.665 Successfully installed at /root/miniconda-4.14.0. Please activate with 'source setup/activateXXX.sh' in every terminal where you want to use conda
#10 0.666 setup/checks/check_for_conda.sh: line 2: conda: command not found
#10 0.668 For conda, found , expected 4.14.0, run 'bash setup/setup_conda.sh <platform>' to get the correct version
#10 0.668 Or install manually after downloading from https://repo.anaconda.com/miniconda/
#10 0.668 setup/checks/check_for_conda.sh: line 4: [: ==: unary operator expected
#10 0.668 Installing using conda now
#10 0.668 setup/setup.sh: line 11: conda: command not found
------
executor failed running [/bin/sh -c /bin/bash -c "cd e-mission-server && cat setup/export_versions.sh && source setup/setup_conda.sh Linux-x86_64 && source setup/setup.sh"]: exit code: 127
ERROR: Service 'notebook-server' failed to build : Build failed
```
</details>

The "correct" way to implement this fix, I think, is to commit updates to [e-mission-server/setup/export_versions.sh](https://github.com/e-mission/e-mission-server/blob/07e01eab5accb9b4f8204ecab86c0c79caf3dfa9/setup/export_versions.sh) and [e-mission-server/setup/environment36.notebook.additions.yml](https://github.com/e-mission/e-mission-server/blob/07e01eab5accb9b4f8204ecab86c0c79caf3dfa9/setup/environment36.notebook.additions.yml) with updated version numbers. Then the image should be built and pushed to docker hub with a new version number that can be used by our `Dockerfile` here. I don't know if this is the desired solution at the time, and we have a waiting [PR](https://github.com/e-mission/e-mission-server/pull/875) that changes the way server images are built.

To implement the fix only changing files in this repo and using the current versions of the server image that are available on docker hub, I have the ugly solution seen here: #53

**Comment by aGuttman** [comment]
Made a new PR that has the changes to the server repo if we want to build and push a new image.
https://github.com/e-mission/e-mission-server/pull/876

**Comment by shankari** [comment]
@aGuttman you are right that the correct way to implement this fix is to change the versions in the `e-mission-server` setup directories.

However, this change does not depend on the waiting PR. As long as containers depend on the `emission/e-mission-server.dev.server-only` image, they will re-check out the code and re-setup when started.

Reviewing https://github.com/e-mission/e-mission-server/pull/876 now...

**Comment by shankari** [comment]
@aGuttman when we deployed an image with these changes on AWS, we ran into some false positives on the vulnerability scan
https://github.com/e-mission/em-public-dashboard/commit/170a7d48435fb79859a473f9303bf21ca44d988f

Can you see whether the `conda -t` and `conda -p` after adding the auxiliary packages (`conda update...`) fixes these false positives?

If so, can you add the same commands (`conda -t` and `conda -p`) to the public dashboard Dockerfile?
If not, can add the `rm -rf` line to the `Dockerfile` in this repo?

Once that is done, I think we can go ahead and merge this since we have already been testing this using the public dashboard image.

**Comment by shankari** [comment]
@aGuttman when we deployed an image with these changes on AWS, we ran into some false positives on the vulnerability scan
https://github.com/e-mission/em-public-dashboard/commit/170a7d48435fb79859a473f9303bf21ca44d988f

Can you see whether the `conda -t` and `conda -p` after adding the auxiliary packages (`conda update...`) fixes these false positives?

If so, can you add the same commands (`conda -t` and `conda -p`) to the public dashboard Dockerfile?
If not, can add the `rm -rf` line to the `Dockerfile` in this repo?

Once that is done, I think we can go ahead and merge this since we have already been testing this using the public dashboard image.

**Comment by aGuttman** [comment]
Removing the `rm -rf` line and replacing with `conda clean -t` and `conda clean -p` does not clear up the false positives from package tests.

`conda clean -t` and `conda clean -p` in `e-mission-server` replaced with `find /root/miniconda-*/pkgs -wholename \*info/test\* -type d | xargs rm -rf` as suggested above. Image builds, container runs, package test files are removed as desired.

**Comment by shankari** [comment]
I think you should keep the `conda clean -t` etc - that removes the tarballs for the install
https://docs.conda.io/projects/conda/en/latest/commands/clean.html

```
-p, --packages

    Remove unused packages from writable package caches. WARNING: This does not check for packages installed using symlinks back to the package cache.
-t, --tarballs

    Remove cached package tarballs.
```

**Comment by aGuttman** [comment]
Should I add `conda clean-t/-p` to the public dashboard Dockerfile? They don't fix the false positive problem, but are they still useful to have here?

**Comment by shankari** [comment]
Yup! I think that both the Dockerfiles should cleanup the install in the same way.
However, I think that the changes to the public dashboard are more substantive.

Notably, we should change it to use our pre-built server docker image instead of using `./clone_server.sh`
And we should create a CI that will build the dashboard image when there is a change.

Let's track the dashboard Dockerfile cleanup + CI in a separate PR.

**Comment by shankari** [comment]
Dashboard dockerfile cleanup + CI + switch for including/excluding test users tracked in https://github.com/e-mission/e-mission-docs/issues/809

This has been fixed in https://github.com/e-mission/e-mission-server/pull/876 and https://github.com/e-mission/e-mission-server/pull/878
